### PR TITLE
Add simple support for Micrometer

### LINKED ISSUE #2344: Micrometer support

### PR DESCRIPTION
**Pull request by tjquinno**
Resolves #2344 

Adds `MicrometerSupport` in SE and a CDI extension and related interceptors, etc. for MP support.

### CODE REVIEW

**Comment by tjquinno** [comment]
I have optimistically marked this with the 2.2.0 milestone. We'll see if I complete the MP work and the SE and MP doc and examples and can do adequate testing. 

We might consider this an "experimental" feature for now??

File: common/common/src/main/java/io/helidon/common/FeatureCatalog.java [hunk 1]
```diff
@@ -116,6 +116,10 @@
                     .nativeSupported(false)
                     .flavor(HelidonFlavor.SE)
         );
+        addSe("io.helidon.metrics.micrometer",
```
Comment (tomas-langer) [inline common/common/src/main/java/io/helidon/common/FeatureCatalog.java:119]: Did you check native image support?

File: micrometer/metrics/pom.xml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <artifactId>helidon-project</artifactId>
+        <groupId>io.helidon</groupId>
+        <version>2.1.1-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.helidon.microprofile.metrics</groupId>
```
Comment (tomas-langer) [inline micrometer/metrics/pom.xml:13]: The group ID an artifact ID do not match the location. Also this module seems to be empty?

**Review by tomas-langer — CHANGES_REQUESTED** [review]

File: docs/se/metrics/03_micrometer.adoc [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+Each meter registry has its own way of producing output.
+Write your handler so that it has a reference to the meter registry it should use and so
+its `accept` method sets the payload in the HTTP response using the registry's mechanism for creating output.
+
+* Write a `Function` which accepts a `ServerRequest` and returns an `Optional<Handler>` +
++
+In general, your function looks at the request--the `Content-Type`, query parameters, etc.--to
+decide whether your handler should respond to the request.
+If so, your function should instantiate your `Handler` and return an `Optional.of(theInstance)`;
+otherwise, return your function should return `Optional.empty()`. +
```
Comment (ljamen) [inline docs/se/metrics/03_micrometer.adoc:189]: otherwise, your function should return ... (extra 'return'?)

**Review by ljamen — COMMENTED** [review]
Just some minor changes/typos.

File: micrometer/metrics/pom.xml (hunk 2 above)
Comment (tjquinno) [reply micrometer/metrics/pom.xml:13]: The PR is a draft and the MP parts are not pushed yet. I pushed this file prematurely.

Comment (tjquinno) [reply micrometer/metrics/pom.xml]: I incorrectly added this.

File: docs/se/metrics/03_micrometer.adoc [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+///////////////////////////////////////////////////////////////////////////////
+
+= Micrometer Metrics
+:description: Helidon Micrometer metrics
+:keywords: micrometer, helidon, metrics
+:javadoc-base-url-api: {javadoc-base-url}io.helidon.metrics.micrometer/io/helidon/metrics/micrometer
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc:22]: fixed

File: docs/se/metrics/03_micrometer.adoc [hunk 5]
```diff
@@ -0,0 +1,226 @@
...
+To use other meter registry types, you will need to add dependencies for them to your `pom.xml`.
+
+== Using Micrometer in Your Application
+You need to make two types of changes to your application to use Helidon's Micrometer support:
+
+1. Register an instance of link:{javadoc-base-url-api}/MicrometerSupport.html[`MicrometerSupport`] with the web server.
+2. Modify your application service code to use the meter registry which `MicrometerSupport` manages to create
+and update meters.
+
+=== Register an Instance of `MicrometerSupport` with the Web Werver
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc:57]: fixed

File: docs/se/metrics/03_micrometer.adoc [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+                        MicrometerSupport.BuiltInRegistryType.PROMETHEUS,
+                        myPrometheusConfig)
+                .build();
+----
+<1> Create the meter registry configuration however you need.
+<2> Enroll the `PROMETHEUS` built-in registry type with your custom configuration.
+
+
+==== Using Configuration
+To use configuration to control the selection and behavior of Helidon's built-in Micrometer meter registries
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc:146]: fixed

File: docs/se/metrics/03_micrometer.adoc (hunk 3 above)
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc:189]: fixed

File: docs/se/metrics/03_micrometer.adoc [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+----
+metrics:
+  micrometer:
+    builtin-registries:
+      - type: prometheus
+        prefix: myPrefix
+----
+
+Which configuration keys are valid in each `builtin-registries` entry depends on the type of Micrometer meter registry.
+Refer to the documentation for the meter registry you want to configure to find out what items apply to that registry
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc:169]: No link. There are many possible other Micrometer registry implementations a developer might use, each with its own peculiarities of configuration and doc link. All we can do here is highlight that the developer has to find out what config can be set and reinforce that Helidon cannot validate it.

File: metrics/micrometer/src/main/java/io/helidon/metrics/micrometer/package-info.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Basic Micrometer support for Helidon SE apps.
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply metrics/micrometer/src/main/java/io/helidon/metrics/micrometer/package-info.java:18]: fixed

File: common/common/src/main/java/io/helidon/common/FeatureCatalog.java (hunk 1 above)
Comment (tjquinno) [reply common/common/src/main/java/io/helidon/common/FeatureCatalog.java:119]: It worked for me in SE which is all the PR has so far. Did you see something that worried you about native image support?

File: docs/se/metrics/03_micrometer.adoc [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+///////////////////////////////////////////////////////////////////////////////
+
+= Micrometer Metrics
+:description: Helidon Micrometer metrics
+:keywords: micrometer, helidon, metrics
+:javadoc-base-url-api: {javadoc-base-url}io.helidon.metrics.micrometer/io/helidon/metrics/micrometer
+
+Helidon SE simplifies how you can use Micrometer for application-specific metrics:
+
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc:26]: fixed

File: docs/se/metrics/03_micrometer.adoc [hunk 10]
```diff
@@ -0,0 +1,226 @@
...
+        request.next();
+    }
+}
+----
+<1> Use the Micrometer meter registry to create the request counter.
+<2> Add routing for any request to invoke the method which counts requests by updating the counter.
+<3> Update the counter and delegate the rest of the request processing to the next handler in the chain.
+
+The example above enrolls the built-in Prometheus meter registry with the default Prometheus registry configuration.
+You can change the default set-up for built-in registries, and you can enroll other meter registries your application
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc]: fixed

File: docs/se/metrics/03_micrometer.adoc [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+.Enroll Prometheus built-in meter registry with non-default configuration
+----
+metrics:
+  micrometer:
+    builtin-registries:
+      - type: prometheus
+        prefix: myPrefix
+----
+
+Which configuration keys are valid in each `builtin-registries` entry depends on the type of Micrometer meter registry.
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc]: fixed

File: docs/se/metrics/03_micrometer.adoc [hunk 12]
```diff
@@ -0,0 +1,226 @@
...
+Refer to the documentation for the meter registry you want to configure to find out what items apply to that registry
+type.
+
+Helidon does not validate the configuration keys you specify against the items defined by the corresponding
+meter registry configuration class.
+
+=== Enrolling other Micrometer meter registries
+To create your own registries and enroll them with `MicrometerSupport`, you need to:
+
```
[the remark this answers is not in the document]
Comment (tjquinno) [reply docs/se/metrics/03_micrometer.adoc:179]: fixed

**Comment by tjquinno** [comment]
There is significant overlap between the Micrometer and the MP-inspired metrics support, in SE and MP. As a result, the most recent commits do away with the separate module for Micrometer support in SE and abstracts common behavior into a base class which `MetricsSupport` and `MicrometerSupport` both extend and moves `MicrometerSupport` into the  `metrics/metrics` module.

The Micrometer-related dependencies are marked as `optional` in the pom so only SE apps which actually use Micrometer support will trigger loading of the Micrometer-related classes.

Now on to the same approach for MP.

**Pull request closed without merging**